stratosphere/ncm/source/lr_ilocationresolver.hpp

```diff
@@ -0,0 +1,139 @@
+                this->html_docs_redirector.ClearRedirections(excluding_ids, num_ids);
+                this->legal_info_redirector.ClearRedirections(excluding_ids, num_ids);
+            }
+        public:
+            virtual Result ResolveProgramPath(sf::Out<Path> out, ncm::ProgramId id) = 0;
+            virtual Result RedirectProgramPath(const Path &path, ncm::ProgramId id) = 0;
+            virtual Result ResolveApplicationControlPath(sf::Out<Path> out, ncm::ProgramId id) = 0;
+            virtual Result ResolveApplicationHtmlDocumentPath(sf::Out<Path> out, ncm::ProgramId id) = 0;
+            virtual Result ResolveDataPath(sf::Out<Path> out, ncm::ProgramId id) = 0;
+            virtual Result RedirectApplicationControlPathDeprecated(const Path &path, ncm::ProgramId id) = 0;
+            virtual Result RedirectApplicationControlPath(const Path &path, ncm::ProgramId id, ncm::ProgramId owner_id) = 0;
+            virtual Result RedirectApplicationHtmlDocumentPathDeprecated(const Path &path, ncm::ProgramId id) = 0;
+            virtual Result RedirectApplicationHtmlDocumentPath(const Path &path, ncm::ProgramId id, ncm::ProgramId owner_id) = 0;
+            virtual Result ResolveApplicationLegalInformationPath(sf::Out<Path> out, ncm::ProgramId id) = 0;
+            virtual Result RedirectApplicationLegalInformationPathDeprecated(const Path &path, ncm::ProgramId id) = 0;
+            virtual Result RedirectApplicationLegalInformationPath(const Path &path, ncm::ProgramId id, ncm::ProgramId owner_id) = 0;
+            virtual Result Refresh() = 0;
+            virtual Result RedirectApplicationProgramPathDeprecated(const Path &path, ncm::ProgramId id) = 0;
+            virtual Result RedirectApplicationProgramPath(const Path &path, ncm::ProgramId id, ncm::ProgramId owner_id) = 0;
+            virtual Result ClearApplicationRedirectionDeprecated() = 0;
+            virtual Result ClearApplicationRedirection(const sf::InArray<ncm::ProgramId> &excluding_ids) = 0;
+            virtual Result EraseProgramRedirection(ncm::ProgramId id) = 0;
+            virtual Result EraseApplicationControlRedirection(ncm::ProgramId id) = 0;
+            virtual Result EraseApplicationHtmlDocumentRedirection(ncm::ProgramId id) = 0;
+            virtual Result EraseApplicationLegalInformationRedirection(ncm::ProgramId id) = 0;
+            virtual Result ResolveProgramPathForDebug(sf::Out<Path> out, ncm::ProgramId id) = 0;
+            virtual Result RedirectProgramPathForDebug(const Path &path, ncm::ProgramId id) = 0;
+            virtual Result RedirectApplicationProgramPathForDebugDeprecated(const Path &path, ncm::ProgramId id) = 0;
+            virtual Result RedirectApplicationProgramPathForDebug(const Path &path, ncm::ProgramId id, ncm::ProgramId owner_id) = 0;
+            virtual Result EraseProgramRedirectionForDebug(ncm::ProgramId id) = 0;
+        public:
+            DEFINE_SERVICE_DISPATCH_TABLE {
```

Comment:
I think the virtual interface should not have a SERVICE_DISPATCH_TABLE.

Not putting it here will allow reusing this class for client code, which I would like to do very much.

Reply:
If it's a requirement for libstratosphere, then add a class ILocationResolverInterface which inherits from ILocationResolver + has the table.